src/h3lib/lib/localij.c

```diff
@@ -542,3 +540,109 @@ int H3_EXPORT(h3Distance)(H3Index origin, H3Index h3) {
 
     return ijkDistance(&originIjk, &h3Ijk);
 }
+
+/**
+ * Number of indexes in a line from the start index to the end index,
+ * to be used for allocating memory. Returns a negative number if the
+ * line cannot be computed.
+ *
+ * @param start Start index of the line
+ * @param end End index of the line
+ * @return Size of the line, or a negative number if the line cannot
+ * be computed.
+ */
+int H3_EXPORT(h3LineSize)(H3Index start, H3Index end) {
+    int distance = H3_EXPORT(h3Distance)(start, end);
+    return distance >= 0 ? distance + 1 : distance;
+}
+
+/**
+ * Given cube coords as floats, round to valid integer coordinates. Algorithm
+ * from https://www.redblobgames.com/grids/hexagons/#rounding
+ * @param i   Floating-point I coord
+ * @param j   Floating-point J coord
+ * @param k   Floating-point K coord
+ * @param ijk IJK coord struct, modified in place
+ */
+static void cubeRound(float i, float j, float k, CoordIJK* ijk) {
+    int ri = round(i);
+    int rj = round(j);
+    int rk = round(k);
+
+    float iDiff = fabs((float)ri - i);
+    float jDiff = fabs((float)rj - j);
+    float kDiff = fabs((float)rk - k);
+
+    // Round, maintaining valid cube coords
+    if (iDiff > jDiff && iDiff > kDiff) {
+        ri = -rj - rk;
+    } else if (jDiff > kDiff) {
+        rj = -ri - rk;
+    } else {
+        rk = -ri - rj;
+    }
+
+    ijk->i = ri;
+    ijk->j = rj;
+    ijk->k = rk;
+}
+
+/**
+ * Given two H3 indexes, return the line of indexes between them (inclusive).
+ *
+ * This function may fail to find the line between two indexes, for
+ * example if they are very far apart. It may also fail when finding
+ * distances for indexes on opposite sides of a pentagon.
+ *
+ * Notes:
+ *
+ *  - The specific output of this function should not be considered stable
+ *    across library versions. The only guarantees the library provides are
+ *    that the line length will be `h3Distance(start, end) + 1` and that
+ *    every index in the line will be a neighbor of the preceding index.
+ *  - Lines are drawn in grid space, and may not correspond exactly to either
+ *    Cartesian lines or great arcs.
+ *
+ * @param start Start index of the line
+ * @param end End index of the line
+ * @param out Output array, which must be of size h3LineSize(start, end)
+ * @return 0 on success, or another value on failure.
+ */
+int H3_EXPORT(h3Line)(H3Index start, H3Index end, H3Index* out) {
+    int distance = H3_EXPORT(h3Distance)(start, end);
+    // Early exit if we can't calculate the line
+    if (distance < 0) {
+        return distance;
+    }
+
+    // Get IJK coords for the start and end. We've already confirmed
+    // that these can be calculated with the distance check above.
+    CoordIJK startIjk = {0};
+    CoordIJK endIjk = {0};
+
+    // Convert H3 addresses to IJK coords
+    h3ToLocalIjk(start, start, &startIjk);
```

Comment:
Just a comment on our `experimentalH3ToLocalIj` transforms: It's unfortunate that the coordinates are not offset by the origin coordinates in the calculation, then you would only have to call `h3ToLocalIjk(start, end, &endIjk)` because `h3ToLocalIjk(start, start, &startIjk)` would've been defined as `(0, 0, 0)`. A bit less computation and the relative nature of the coordinate system would be more obvious to users.

@isaacbrodsky thoughts?

Reply:
FWIW, I don't think it's any less computation, since you'd still be calculating `h3ToLocalIjk(start, start, &startIjk)` under the hood in order to get the coordinates for offset. In fact, it's probably worse, because calling `h3ToLocalIjk` repeatedly will perform the calculation of the offset every time - the only time you save work is when the `origin` is the same as the index you want to get coords for.

Reply:
`h3ToLocalIjk` includes several branches and loops. I'd assume that three subtractions would be minimal in comparison to the rest of the function, and it would let you call that function once instead of twice *and* produce coordinates that are simpler to reason about.

Reply:
But you'd need to get the coordinates to subtract, which is effectively an `h3ToLocalIjk` call on your origin, right? Right now, the origin is just used to get the base cell, and then the IJK is calculated for the target; if you offset by the origin, you'd need to calculate the IJK for the origin, the IJK for the target, and then translate by the origin. It's the additional IJK calculation that adds the perf impact, not the translation.

I could be wrong here, but pretty sure that in a case like `h3Line`, where you are finding the coordinates for indexes `1, 2, 3, ...N`, you'd have `N - 1` calls instead of the `N` we have now, but you'd actually be doing `(N - 1) * 2` IJK calculations, instead of the `N` we do now.